sdk/servicebus/Azure.Messaging.ServiceBus/api/Azure.Messaging.ServiceBus.netstandard2.0.cs

```diff
@@ -777,6 +780,31 @@ public sealed partial class TrueRuleFilter : Azure.Messaging.ServiceBus.Manageme
         public override string ToString() { throw null; }
     }
 }
+namespace Azure.Messaging.ServiceBus.Plugins
```

Comment:
I'm not opposed to Policy. The common use case for apps would be using Policies/Plugins built by someone else, so Plugin is nice in that sense.

Reply:
Prior art in track one aside, I don't know that `Policy` would be the way that I go here; the connotation around policies are typically associated with security in my experience.  If I'm being honest, I'm not a fan of our use in the HTTP pipeline space either.  I would be concerned with confusion here.   

Personally, I believe that we'll want to keep the existing terminology, but if we decide to do otherwise, I'd propose something closer to `MessageTransformer` or `MessageProcessor` or similar to keep the association with its context of inspecting/adjusting a message.

Reply:
It is definitely not a policy. An extension would hint that something new is added, which is not quite true as we get and return a message. A plugin is a well-established nomenclature.

Reply:
@jsquire - what are you going to call the thing that sits in the middle of the AMQP stream fiddling with messages?  My thought is the name should match that.  For HTTP, we call that `HttpPipelinePolicy`.

Reply:
@tg-msft note that it's not operating at the AMQP level but a level higher. HTTP pipeline policies you're referring to do operate at the HTTP level.

Reply:
Understood - this class is not operating at the AMQP level so I was suggesting `MessagePolicy` as a suffix.  If we add an `AmqpPipeline` at some point with `AmqpPipelinePolicy` classes, I would like the names for similar concepts to line up.  I also want to give this a name that will work across Track 2 libraries, whether they're AMQP or HTTP, so we could possibly add something similar to Storage Queues someday.

To Josh's point, I don't think most folks are going to be writing their own `HttpPipelinePolicy`s either.  To Jesse's suggestions, I don't like `Transformer` because it might not actually change anything and `Processor` doesn't feel so much better than `Policy` to deviate from a naming pattern we're already locked into.

Maybe a better question to ask is how this differs from an `HttpPipelinePolicy` other than the type of message being passed through?  It's obviously not chaining policies, but are there other major differences in the goals between the two?

Reply:
> Maybe a better question to ask is how this differs from an `HttpPipelinePolicy` other than the type of message being passed through? 

That's a good question. 

> It's obviously not chaining policies, but are there other major differences in the goals between the two?

From my limited exposure to `HttpPipelinePolicy` first thing that is obvious is the design intention behind it. It's not designed to support incoming and outgoing pipelines for the payload but more of a lower control/cross-cutting concerns mechanism. It has a general `ProcessAsync(HttpMessage)` where HttpMessage is a non-deterministic payload that is not associated with any specific service. It could be Queue Storage message, CosmosDB record(s), or anything else that communicates via HTTP. Plugins know exactly what they operate on - Service Bus message, it's body and metadata. And if necessary, it can even operate on the user-level concrete type by converting the payload back into a POCO.

Then there's the fact that these policies are interconnected, exposing `ProcessNextAsync()` with the entire policy chain. One policy can cut out other policies. Plugins are isolated and don't interact with each other, just the payload.

And last, but not least, the design of `HttpPipelinePolicy` does not differentiate between incoming and outgoing traffic. It all goes through `ProcessAsync()`. It entirely based on the HTTP pipeline with Request and Response objects as the primary target, where the payload is less of a significance from the policy perspective. Those policies are truly policies that enforce things such as logging, authentication, retries, networking, distributed tracing. It's about requests and responses over HTTP, never the data itself.

The goal of a plugin is to be self-isolated, seamlessly be added alongside with other plugins, and strictly operate on the content/payload level.

Reply:
I've asked the community (your customers) what do they think of the name options. Here's the [on-going poll](https://twitter.com/sfeldman/status/1273845277032509442).

Reply:
For me it's a no-brainer - What we are introducing here are plug-ins; features that are not part of the original SDK but we can bring our own. This is very similar where you have Visual Studio (~SDK) where I can extend it by installing a plugin.

Here's why I don't like names like these:
- `Policy` is to enforce things such as security, but only the minority of plugins are doing that so it's too narrow
- `Transformer` is to transform message payloads on read/write, but only the minority of plugins are doing that so it's too narrow
- `Processor` processes things but I don't see why this would be useful here.

I would highly suggest to stick with plugin as everybody is familiar with this concept and I don't see any reason why it's not good?

Reply:
> @jsquire - what are you going to call the thing that sits in the middle of the AMQP stream fiddling with messages? My thought is the name should match that. For HTTP, we call that `HttpPipelinePolicy`.

I would disagree on that point.  I was intentionally trying to keep the names distinct to avoid confusion because both the implementation and purpose is quite different.   Conceptually, a `Plugin` is operating in the context of the application and is something that we would want to steer folks towards for the large majority of their needs.

What I was calling a `MessageFormatter` is operating in the context of a transport and, thus, requires low-level knowledge of the active protocol - AMQP, for example.  This would be something that we expect only developers with specialized needs around interop with other blackbox systems would need to dig into.  It definitely has a much smaller and more focused target audience than `Plugins`.   That said, it turns out that it is very difficult to get to this level without leaking details of your underlying transport dependency or creating an internal model hierarchy to mirror that.   We're currently trying our hand at a different approach, so this is not likely a concept that will make it in.

While I agree that consistency is a worthy goal, I feel that the audience for `Plugins` being larger than those that would want to customize the HTTP pipeline.   I'd prefer to take a ding on consistency in favor of a name that has more meaning in the existing ecosystem and which doesn't run the risk of confusion due to the prevalence of "security policies" in the services and application space.

Reply:
> I've asked the community (your customers) what do they think of the name options. Here's the [on-going poll](https://twitter.com/sfeldman/status/1273845277032509442).

Thank you, Sean, for raising awareness.  Community feedback for these discussions is very much appreciated.